File: rust/koru/src/vocab.rs
// SPDX-License-Identifier: MIT

//! Braam's vocabulary in Rust: the fifteen names, `Result`, and the aliases
//! Rust spells differently. Not a second `Error` — koru-sys's, re-exported.
//! The `?` conversions are in koru-sys, where the orphan rule puts them.
//! doc/Notes.md has the reasoning.

pub use koru_sys::error::{Errno, Error, KINDS, Kind};

/// Braam's `Result<T, E = Error>`. `TRY` and `CO_TRY` are Rust's `?`.
pub type Result<T, E = Error> = core::result::Result<T, E>;

/// Braam's `Str`: a non-owning view of UTF-8 bytes.
pub type Str<'a> = &'a str;

/// Braam's `Span<const T>`. `String`, `Option` and the integer names are
/// Rust's own, so these three are the whole alias list.
pub type Span<'a, T> = &'a [T];

/// Braam's `Span<T>`, the mutable one.
pub type SpanMut<'a, T> = &'a mut [T];

#[cfg(test)]
mod tests {
    use super::*;
    use koru_sys::EnterError;
    use koru_sys::error::{EPIPE, KORU_ERRNOS, kind_of};
    use std::collections::{HashMap, HashSet};
    use std::io;

    /// Takes koru-sys's `Error` by its own path: a second type defined here
    /// would not compile.
    fn as_sys(e: koru_sys::error::Error) -> koru_sys::error::Kind {
        e.kind()
    }

    #[test]
    fn the_vocabulary_is_koru_syss_types_re_exported() {
        let e: Error = Errno(22).into();
        assert_eq!(as_sys(e), Kind::Invalid);
    }

    /// The done test, over `KORU_ERRNOS` unchanged.
    #[test]
    fn every_errno_maps_to_exactly_one_name_and_back_to_its_raw_value() {
        let mut seen: HashMap<i32, Kind> = HashMap::new();
        for d in KORU_ERRNOS {
            let e = Error::from(d.errno);
            assert_eq!(e.kind(), d.kind, "{} took the wrong name", d.name);
            assert_eq!(e.raw(), d.errno, "{} lost its raw errno", d.name);
            assert!(e.is(d.kind));
            assert_eq!(e, d.kind, "{} fails the bare-name comparison", d.name);
            assert_eq!(kind_of(d.errno), d.kind);
            assert!(
                seen.insert(d.errno.0, d.kind).is_none(),
                "{} appears twice",
                d.name
            );
        }
        assert_eq!(seen.len(), KORU_ERRNOS.len());
    }

    /// `?` from each error type a koru program meets. A missing `From` fails
    /// this at compile time, which is why the loop asserts so little.
    #[test]
    fn the_question_mark_converts_every_errno_koru_can_produce() {
        fn from_errno(e: Errno) -> Result<()> {
            Err(e)?;
            Ok(())
        }
        fn from_io(e: io::Error) -> Result<()> {
            Err(e)?;
            Ok(())
        }
        fn from_enter(e: EnterError) -> Result<()> {
            Err(e)?;
            Ok(())
        }

        for d in KORU_ERRNOS {
            let want = Error::from(d.errno);
            assert_eq!(from_errno(d.errno).unwrap_err(), want);
            assert_eq!(from_io(d.errno.as_io()).unwrap_err(), want);
            let entered = EnterError {
                errno: d.errno,
                progress: Default::default(),
            };
            assert_eq!(from_enter(entered).unwrap_err(), want);
        }
    }

    /// Back to a raw errno through an ordinary Rust caller.
    #[test]
    fn every_error_round_trips_through_io_error() {
        for d in KORU_ERRNOS {
            let e = Error::from(d.errno);
            let io: io::Error = e.into();
            assert_eq!(io.raw_os_error(), Some(d.errno.0), "{}", d.name);
            assert_eq!(Error::from(io), e, "{}", d.name);
        }
    }

    /// A name no errno reaches is a name no program can be handed. Since T21
    /// there are none: `Closed` arrives as EPIPE from a write, and end of file
    /// synthesises it with no errno at all.
    #[test]
    fn every_name_is_reachable_from_the_errno_table() {
        let mapped: HashSet<Kind> = KORU_ERRNOS.iter().map(|d| d.kind).collect();
        let orphans: Vec<Kind> = KINDS
            .iter()
            .copied()
            .filter(|k| !mapped.contains(k))
            .collect();
        assert_eq!(orphans, [] as [Kind; 0], "a name no errno can produce");
        assert_eq!(Error::from(EPIPE).kind(), Kind::Closed);
        assert_eq!(Error::closed().kind(), Kind::Closed);
        assert_eq!(Error::closed().raw(), Errno(0), "end of file has no errno");
    }

    #[test]
    fn the_aliases_spell_braams_signatures() {
        fn first(s: Str<'_>) -> Option<char> {
            s.chars().next()
        }
        fn total(xs: Span<'_, u8>) -> u32 {
            xs.iter().map(|&b| b as u32).sum()
        }
        fn zero(xs: SpanMut<'_, u8>) {
            xs.fill(0);
        }

        let mut buf = [1u8, 2, 3];
        assert_eq!(first("koru"), Some('k'));
        assert_eq!(total(&buf), 6);
        zero(&mut buf);
        assert_eq!(total(&buf), 0);
    }
}

File: cpp/tests/vocab.cpp
// SPDX-License-Identifier: MIT
//
// T44's host half: the vocabulary, the `TRY` family and `Args`. None of it
// touches /dev/koru, so ctest runs it on the host.
//
// The errno table itself is T39's, in unit.cpp; what is new here is that every
// name a program can be handed is reachable from that table, and that `TRY`
// converts each of the error types a koru program meets.

#include "harness.hpp"

#include <koru/args.hpp>
#include <koru/task.hpp>
#include <koru/vocab.hpp>

#include <cerrno>
#include <set>
#include <string>

using namespace koru;

namespace {

// A comma inside a template argument list is a comma to the preprocessor, so
// the two-parameter results get a name of their own before they reach a macro.
using ErrnoVoid = result<void, Errno>;
using EnterVoid = result<void, EnterError>;

/// `TRY` from each error type. A missing `as_error` overload fails this at
/// compile time, which is why the bodies assert so little.
result<int> from_errno_value(Errno e)
{
    TRY_VOID(ErrnoVoid(e));
    return 0;
}

result<int> from_enter(EnterError e)
{
    TRY_VOID(EnterVoid(e));
    return 0;
}

result<int> from_error(Error e)
{
    TRY_VOID(result<void>(e));
    return 0;
}

/// And the value-yielding arm, which must move rather than copy.
result<String> try_moves_its_value()
{
    String s = TRY(result<String>(String("a string long enough not to fit in a small buffer")));
    return s;
}

task<result<int>> co_try_from_errno(Errno e)
{
    CO_TRY_VOID(ErrnoVoid(e));
    co_return 0;
}

task<result<String>> co_try_moves_its_value()
{
    String s = CO_TRY(result<String>(String("another string past the small-buffer size")));
    co_return std::move(s);
}

// Braam's signatures, in Braam's aliases. Never called in anger: a drift in
// one of the four is a compile error.
Option<char> first(Str s)
{
    return s.empty() ? Option<char>() : Option<char>(s.front());
}

u32 total(Span<u8> xs)
{
    u32 n = 0;
    for (u8 b : xs)
        n += b;
    return n;
}

void zero(SpanMut<u8> xs)
{
    for (u8 &b : xs)
        b = 0;
}

} // namespace

CASE(vocab_every_errno_converts_through_try_to_the_same_error)
{
    for (size_t i = 0; i < koru_errnos_len(); i++) {
        const ErrnoDef &d = KORU_ERRNOS[i];
        Error want        = Error::from_errno(d.err);

        result<int> a = from_errno_value(d.err);
        result<int> b = from_enter(EnterError{ d.err, Progress{} });
        result<int> c = from_error(want);
        if (a.ok() || b.ok() || c.ok()) {
            FAILF("%s: TRY_VOID let an error through", d.name);
            continue;
        }
        if (!(a.error() == want) || !(b.error() == want) || !(c.error() == want))
            FAILF("%s: TRY_VOID converted to a different error", d.name);

        // And the same inside a coroutine, which is a different statement.
        result<int> co = sync_wait(co_try_from_errno(d.err));
        if (co.ok() || !(co.error() == want))
            FAILF("%s: CO_TRY_VOID disagrees with TRY_VOID", d.name);
    }
}

CASE(vocab_try_yields_the_value_and_moves_it)
{
    result<String> r = try_moves_its_value();
    CHECK(r.ok());
    CHECK(r.value().size() > 32);

    result<String> c = sync_wait(co_try_moves_its_value());
    CHECK(c.ok());
    CHECK(c.value().size() > 32);
}

/// A name no errno reaches is a name no program can be handed. Since T21 there
/// are none: `Closed` arrives as EPIPE from a write, and end of file
/// synthesises it with no errno at all.
CASE(vocab_every_name_is_reachable_from_the_errno_table)
{
    std::set<int> mapped;
    for (size_t i = 0; i < koru_errnos_len(); i++)
        mapped.insert(int(KORU_ERRNOS[i].kind));

    for (size_t i = 0; i < 15; i++)
        if (!mapped.count(int(KINDS[i])))
            FAILF("a name no errno can produce: %s", kind_name(KINDS[i]));

    CHECK(Error::from_errno(Errno(EPIPE)).kind() == Kind::Closed);
    CHECK(Error::closed().kind() == Kind::Closed);
    CHECK(Error::closed().raw() == Errno(0)); // end of file has no errno
}

CASE(vocab_the_aliases_spell_braams_signatures)
{
    u8 buf[3] = { 1, 2, 3 };
    CHECK(first("koru").value() == 'k');
    CHECK(!first("").has_value());
    CHECK_EQ(total(Span<u8>(buf, 3)), 6);
    zero(SpanMut<u8>(buf, 3));
    CHECK_EQ(total(Span<u8>(buf, 3)), 0);
}

// ---------------------------------------------------------------------------
// Args
// ---------------------------------------------------------------------------

namespace {

Args three()
{
    return Args(std::vector<String>{ "hello", "koru", "world" });
}

} // namespace

CASE(args_name_is_the_first_and_tail_drops_it)
{
    Args a = three();
    CHECK_EQ(a.size(), 3);
    CHECK(a.name() == "hello");
    CHECK(a[1] == "koru");

    Args t = a.tail();
    CHECK_EQ(t.size(), 2);
    CHECK(t.name() == "koru");
    CHECK(t[0] == "koru");
    CHECK_EQ(a.size(), 3); // tail shares the vector rather than consuming it
}

CASE(args_an_empty_vector_has_an_empty_name_and_a_tail_that_stays_empty)
{
    Args a;
    CHECK(a.empty());
    CHECK(a.name() == "");
    CHECK(a.tail().empty());
    CHECK(a.tail().tail().empty());
}

CASE(args_tail_past_the_end_is_empty_rather_than_a_fault)
{
    Args a = three().tail().tail().tail().tail();
    CHECK(a.empty());
    CHECK(a.name() == "");
}

CASE(args_skip_is_tail_repeated_and_saturates)
{
    Args a = three();
    CHECK_EQ(a.skip(0).size(), 3);
    CHECK(a.skip(2).name() == "world");
    CHECK(a.skip(3).empty());
    CHECK(a.skip(SIZE_MAX).empty());
}
